File: algorithms/N/Nth Digit/Nth Digit.py
class Solution:
    def findNthDigit(self, n: int) -> int:
        low, high = 1, 0
        place = 1
        
        sum_ = 0
        
        while True:
            high = 10**place -1
            digits_sum = (high - low +1 ) * place
            
            if sum_<= n <= digits_sum + sum_:
                break
            else:
                low = high+1
                sum_+= digits_sum
            place += 1
        
        remaining_digits = n - sum_
        
        position = (remaining_digits-1)//place 
        index_ =  remaining_digits - position*place -1
        
        
        number = low + position
        # return str(number)[index_]
        count = 0
        index_ = place - 1 - index_
        l = []
        # 
        while number:
            if count == index_:
                return number%10
            number = number//10
            count+=1

File: scripts/algorithms/N/Nth Digit/Nth Digit.cpp
class Solution {
public:
    #define ll long long 
    int findNthDigit(int n) {
        if(n<=9) return n;
        ll i=1;
        ll sum=0;
        while(true){
            ll a=(9*pow(10,i-1))*i;
            if(n-a<0){
                break;
            } 
            else{
                sum+=9*pow(10,i-1);
                n-=a;
            }
            i++;
        }
        
        ll x=0;        
        ll mod=n%i;
        n=(n/i);
        n+=sum;
        if(mod==0) return n%10;
        else{
            n++;
            ll x=i-mod;
            while(x--){
                n/=10;
            }
        }
        return n%10;
    }
};
